src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const copyState = { ...state };
+  const stateArray = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(copyState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (copyState.hasOwnProperty(key)) {
+            delete copyState[key];
+          }
+        }
+        break;
+
+      case 'clear':
+        for (const key in copyState) {
+          delete copyState[key];
+        }
+        break;
+
+      default:
+        throw new Error('Unknown action type');
+    }
+    stateArray.push(Object.assign({}, copyState));
```

Comment:
```suggestion
    stateArray.push({ ...copyState });
```
you can also spread the copyState object - I think it looks better that way